### ISSUE TITLE
Implementation of `foldRight` and `foldLeft` via `foldMap` is wrong in chapter 10?

### ISSUE THREAD

**Issue by Haidar0096** [fpinscala/fpinscala#648]
In chapter 10 monoids, the implementation of `foldRight` and `foldLeft` via `foldMap` is giving opposite results to whats expected,
the reason is that the book is using the wrong monoids for `foldRight` and `foldLeft`, for `foldRight` it should be the `dual` of the `endoMonoid` defined in the book, and `endoMonoid` for `foldLeft`, I traced the execution and tried a counter-example and I am sure they are wrong, for example: 
```
val list = List("Hello", "world", "!")
    val res4 = foldLeft(list, "")((acc, a) => acc + a) // the normal foldLeft: expected result: HelloWorld!
    val res5 = foldLeftViaFoldMap(list)("")((acc, a) => acc + a) // the wrong implemented one
```
and the result is :
```
res4 = Helloworld!
res5 = !worldHello  // wrong result
```
If we just swap each monoid for its dual in the implementation of `foldLeft` and `foldRight` then this will be solved

**Comment by mpilquist** [comment]
Thanks, fixed in 3fb16cb.

This was a result of changing `endoMonoid` to be defined as `f andThen g` in the second edition, instead of `f compose g` like in the first edition.

**Comment by Haidar0096** [comment]
> Thanks, fixed in [3fb16cb](https://github.com/fpinscala/fpinscala/commit/3fb16cbd7efa2127c27b07d375fd1ead6f477b2b).
> 
> This was a result of changing `endoMonoid` to be defined as `f andThen g` in the second edition, instead of `f compose g` like in the first edition.

thank you, I've been hitting my head against a wall since yesterday, I had to revisit the basic definitions of `foldRight` and `foldLeft` to be sure it's not a misunderstanding on my side,

 have a good day :)
